include/mxnet/engine.h

```diff
@@ -120,6 +120,10 @@ class MXNET_API Engine {
   typedef std::function<void(RunContext)> SyncFn;
   /*! \brief Asynchronous operation to pass to engine. */
   typedef std::function<void(RunContext, CallbackOnComplete)> AsyncFn;
+  /*! \brief Synchronous operation (function pointer) to pass to engine. */
+  typedef void (*SyncFnPtr)(RunContext, const std::shared_ptr<void>&);
```

Comment:
why const reference for the shared_ptr?

Reply:
I wanted to use reference to avoid unnecessary copy of shared_ptr which is an expensive operation. I did some simple experiment and observed that passing shared_ptr by value can be 20x slower compared to passing by reference. In addition, in the implementation of PushAsyncPtr/PushSyncPtr functions, when we create a [lambda function](https://github.com/apache/incubator-mxnet/pull/14615/files#diff-86dd2a82be9dffd682cbdd5c13490ac9R76), we captures the param, which is a copy operation. This is why I decide to use reference here. Then using const to prevent it from being modified.